### ISSUE TITLE
Is it possible to extend existing trained Model with new corpus data

### ISSUE THREAD

**Issue by vattisampath** [bshao001/ChatLearner#45]
Hi Bsha001, 

Thanks for the excellent work and repo to begin with AI Chatbots. Really helped a lot in understanding the AI chatbot concepts and implementations. 

I m able to train my custom corpus data with small sets after some good amount of training time. But now I want to expand my training data. So for that I have to update new data files to Corpus folder.
Then model training will start from beginning with all old trained data too which is time taking again.

Is to possible to pass latest or specific model to training script and start training from that checkpoint.
I have seen such implementation here in this Repo @ https://github.com/daniel-kukiela/nmt-chatbot
Can we also have something like this .. ???

And also is the current implementation is bidirectional RNN ?? And I have seen code for nmt implementation. I didnt find any such module imported, are we using it ? As far I understand its a concept but not a module, but just to make sure and to understand how NMT is applied in this repo.

Once again thanks for an amazing work and sharing them with us. Keeping going broo... !!  Cheers .. 

Awaiting a quick reply for my queries ...

Thanks
Sam

**Comment by anne1994** [comment]
@vattisampath are you using a gpu machine to train the small datasets ?

**Comment by vattisampath** [comment]
@anne1994 No, I m using normal Centos7 32GB server with normal CPUs. No GPU available.

**Comment by bshao001** [comment]
@vattisampath If you can manage to maintain the vocabulary untouched, yes, you can continue to the model from the last checkpoint, although the implementation presented in this repo does not include that part of code. You can make an easy change to the script to reach that goal. If the vocab is changed, you have to train the model from scratch.

Concerning the bidirectional RNN, yes, it is also an easy change. You can refer to the NMT official tutorial   for details.

**Comment by vattisampath** [comment]
Thanks @bshao001 
Understood the importance of vocabulary. Yes we need to maintain the same to restart training from checkpoint. 

Regarding bidirectional RNN, can you please in which exact file of your repo, I need to look for that change. It will be helpful to me if you can point me out specific file and function in that file.

And also, got a new doubt in implementing contexts. Lets say we want to converse only in one context. Can we make bot trained to answer within that context only. 
To make it clear for your understanding what context I mean, I m giving a simple scenario.
We have 2 contexts and few questions and answers in both contexts. Both contexts can have same question, but answer will differ based on the context we speak.
Can you train bot like that .. ?? If its possible, where can i make that differentiation in corpus and how.. ?

Please suggest some way around to achieve this. 

Thanks in adavance.

**Comment by bshao001** [comment]
@vattisampath Concerning bidirectional RNN, you can just make changes here at _single_cell function in https://github.com/bshao001/ChatLearner/blob/master/chatbot/modelhelper.py file. Some possible options might be: tf.nn.bidirectional_dynamic_rnn, or the stacked version. Google the web, or ask questions in stackoverflow for details of that, as this is out of the scope here.

Concerning the contextual issue in a chatbot, it is an active research topic, which we are conducting that as well.

**Comment by anne1994** [comment]
@bshao001  is it possible to get accuracy? , How accurate the answer is to the question, based on training?

**Comment by vattisampath** [comment]
@bshao001  Thanks so much... 

@anne1994  I m not preparing huge dataset. Mine is very small dataset. So its not taking more than 6 hours to finish 150 epoch.

**Comment by bshao001** [comment]
@anne1994 We don't use accuracy for NLP tasks. Perplexity is a more common practice, as far as I know. BLEU is also used for machine translation tasks. For other text generation tasks, such as chatbots, article or poem writing, as there are too many ways of responding a question or a statement, explaining a topic or a feeling,  we don't have a good automatic measurement approach. Manual evaluation is commonly used. You can, however, perform some kind of test, as shown in this repository, which is just to double-check that the training was going properly, i.e., did not stuck in a local optima.
